core/src/Streamly/Internal/Data/Fold/Type.hs

```diff
@@ -12,45 +12,115 @@
 -- We can classify stream consumers in the following categories in order of
 -- increasing complexity and power:
 --
--- == Accumulators
+-- == Folds that never terminate
 --
--- These are the simplest folds that never fail and never terminate, they
--- accumulate the input values forever and can always accept new inputs (never
--- terminate) and always have a valid result value.  A
--- 'Streamly.Internal.Data.Fold.sum' operation is an example of an accumulator.
+-- An Accumulator is the simplest type of fold, it never fails and never
+-- terminates. It can always accept more inputs (never terminates) and the
+-- accumulator is always valid.  For example 'Streamly.Internal.Data.Fold.sum'.
 -- Traditional Haskell left folds like 'foldl' are accumulators.
 --
--- We can distribute an input stream to two or more accumulators using a @tee@
--- style composition.  Accumulators cannot be applied on a stream one after the
--- other, which we call a @serial@ append style composition of folds. This is
--- because accumulators never terminate, since the first accumulator in a
--- series will never terminate, the next one will never get to run.
+-- Accumulators can be composed in parallel where we distribute the input
+-- stream to all accumulators.  Since accumulators never terminate they cannot
+-- be appended.
 --
--- == Terminating Folds
+-- An accumulator can be represented as:
 --
--- Terminating folds are accumulators that can terminate. Once a fold
--- terminates it no longer accepts any more inputs.  Terminating folds can be
--- used in a @serial@ append style composition where one fold can be applied
--- after the other on an input stream. We can apply a terminating fold
--- repeatedly on an input stream, splitting the stream and consuming it in
--- fragments.  Terminating folds never fail, therefore, they do not need
--- backtracking.
+-- @
+-- data Fold0 m a b =
+--   forall s. Fold0
+--      (s -> a -> m s) -- step
+--      (m s)           -- initial
+--      (s -> m b)      -- extract
+-- @
+--
+-- This is just a traditional left fold, compare with @foldl@. The driver of
+-- the fold would call @initial@ at the beginning and then keep accumulating
+-- inputs into its result using @step@ and finally extract the result using
+-- @extract@.
+--
+-- == Folds that terminate after one or more input
+--
+-- Terminating folds are accumulators that can terminate, but like accumulators
+-- they do not fail. Once a fold terminates it no longer accepts any more
```

Comment:
```suggestion
-- Terminating folds are accumulators that can terminate. Once a fold terminates it no longer accepts any more
```

`but` is rather confusing. I expected a contradiction. `but unlike` makes more sense.